src/material/stepper/stepper.ts

```diff
@@ -115,8 +117,144 @@ export class MatStep extends CdkStep implements ErrorStateMatcher, AfterContentI
   }
 }
 
+/**
+ * Proxies the public APIs from `MatStepper` to the deprecated `MatHorizontalStepper` and
+ * `MatVerticalStepper`.
+ * @deprecated Use `MatStepper` instead.
+ * @breaking-change 13.0.0
+ * @docs-private
+ */
+@Directive()
+abstract class _MatProxyStepperBase {
```

Comment:
I don't totally follow why you need `_MatProxyStepperBase`- why can't the shim horizontal/vertical steppers just extend `MatStepper`? Is it to avoid repeating the host bindings for ViewEngine compatibility?

Reply:
I wanted to avoid having to repeat the metadata. Also I believe that the template and styles would be inlined multiple times if I repeated them in the metadata of multiple components. That being said, I'm not 100% sure that it's actually possible for people get a hold of a `MatVerticalStepper` or `MatHorizontalStepper` instance after these changes, but they might still have references to the types. Alternatively, I can turn them into interfaces with the same public API, but it may result in a breaking change if they're referenced as a value, rather than a type.

Reply:
What if we just had `MatStepper` and we did e.g.
```ts
/** @deprecated */
export class MatVerticalStepper extends MatStepper { }

@Component({
  ...,
  providers: [{provide: MatVerticalStepper, useExisting: MatStepper}],
})
export class MatStepper {
}
```

The `extends` here is purely to fake `MatVerticalStepper` being the right type while still keeping it a real class that people can import.

Reply:
I already had the `providers`, but another problem is that `MatVerticalStepper` has to be declared after `MatStepper`, because it extends it, but it also can't be provided in the `providers` of `MatStepper` since it hasn't been declared yet. Using `forwardRef` seems to crash the compiler when used inside `provide`. 

I've tweaked the approach to make it more manageable by having `_MatProxyStepperBase` extend `CdkStepper` and declare the Material-specific APIs on top of it. I've also added a couple of tests to make sure that the deprecated symbols work.